File: 프로그래머스/2/12978. 배달/배달.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

vector<vector<pair<int, int>>> edgeList;


int dijkstra(int start, int N, int k) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    vector<int> dist(N + 1, INT_MAX);
    
    pq.push({0, start});
    dist[start] = 0;
    
    int cnt = 0;
    while(!pq.empty()) {
        auto [curCost, cur] = pq.top();
        pq.pop();
        
        // cnt++;
        if (dist[cur] < curCost) continue;
        if (curCost > k) continue;
        
        for (auto [nxt, nxtCost] : edgeList[cur]) {
            int totalCost = curCost + nxtCost;
            if (dist[nxt] < totalCost) continue;
            if (totalCost > k) continue;
            
            dist[nxt] = totalCost;
            pq.push({totalCost, nxt});
        }
    }
    
    for (int i = 1; i <= N; i++) {
        if (dist[i] <= k) cnt++;
    }
    
    return cnt;
}

int solution(int N, vector<vector<int> > road, int K) {
    edgeList.resize(N + 1);
    for (auto r : road) {
       edgeList[r[0]].push_back({r[1], r[2]});
       edgeList[r[1]].push_back({r[0], r[2]});
    }

    return dijkstra(1, N, K);
}

File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    List<int[]>[] edge;
    int N, K;
    
    public int solution(int N, int[][] road, int K) {
        this.N = N;
        this.K = K;
        
        edge = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            edge[i] = new ArrayList<>();
        }
        
        for (int[] r : road) {
            edge[r[0]].add(new int[]{r[1], r[2]});
            edge[r[1]].add(new int[]{r[0], r[2]});
        }
        
        return dijkstra(1);
    }
    
    private int dijkstra(int start) {
        PriorityQueue<int[]> pq = 
            new PriorityQueue<>((o1, o2) -> Integer.compare(o1[1], o2[1]));
        int[] dist = new int[N + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        
        dist[start] = 0;
        pq.offer(new int[]{start, 0});
        
        while(!pq.isEmpty()) {
            int[] cur = pq.poll();
            int node = cur[0];
            int weight = cur[1];
            
            if (dist[node] < weight) continue;
            
            for (int[] next : edge[node]) {
                int total = weight + next[1];
                if (dist[next[0]] > total) {
                    dist[next[0]] = total;
                    pq.offer(new int[]{next[0], total});
                }
            }
        }
        
        int cnt = 0;
        for (int i = 1; i <= N; i++) {
            if (dist[i] <= K) cnt++;
        }
        return cnt;
    }
}